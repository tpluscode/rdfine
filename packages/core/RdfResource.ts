/* eslint-disable camelcase,@typescript-eslint/camelcase,no-dupe-class-members */
import RDF from '@rdfjs/data-model'
import { NamespaceBuilder } from '@rdfjs/namespace'
import { NamedNode, DatasetCore, BlankNode, DefaultGraph, Quad_Graph, Term, Literal } from 'rdf-js'
import cf, { SafeClownface, SingleContextClownface } from 'clownface'
import ResourceFactoryImpl, {
  Constructor,
  Mixin,
  ResourceCreationOptions,
  ResourceFactory,
  ResourceIndexer,
} from './lib/ResourceFactory'
import once from 'once'
import TypeCollectionCtor, { TypeCollection } from './lib/TypeCollection'
import { xsd } from '@tpluscode/rdf-ns-builders'

type ObjectOrFactory<T> = T | ((self: RdfResource) => T)

export type ResourceIdentifier = BlankNode | NamedNode
export type ResourceNode<D extends DatasetCore = DatasetCore> = { dataset: D; term: Term; graph?: NamedNode | DefaultGraph }

export interface GetOptions {
  strict: boolean
}

export interface RdfResource<D extends DatasetCore = DatasetCore> {
  readonly id: ResourceIdentifier
  readonly types: TypeCollection<D>
  readonly _selfGraph: SingleContextClownface<ResourceIdentifier, D>
  readonly _unionGraph: SafeClownface<ResourceIdentifier, D>
  readonly _graphId: Quad_Graph
  readonly _parent?: RdfResource<D>
  readonly isAnonymous: boolean
  hasType (type: string | NamedNode): boolean
  /**
   * Gets the value of a property
   * @param property
   */
  get<T extends RdfResource = RdfResource> (property: string | NamedNode): T
  get<T extends RdfResource = RdfResource> (property: string | NamedNode, options?: GetOptions): T | null
  /**
   * Gets the value of a property and ensures that an array will be returned
   * @param property
   */
  getArray<T extends RdfResource = RdfResource> (property: string | NamedNode, options?: GetOptions): T[]
  /**
   * Gets the property value if it's boolean. Throws if it's not
   * @param property
   */
  getBoolean (property: string | NamedNode, options?: GetOptions): boolean
  /**
   * Gets the property value if it's a string. Throws if it's not
   * @param property
   */
  getString (property: string | NamedNode): string
  getString (property: string | NamedNode, options?: GetOptions): string | null
  /**
   * Gets the property value if it's a number. Throws if it's not
   * @param property
   */
  getNumber (property: string | NamedNode): number | null
  getNumber (property: string | NamedNode, options?: GetOptions): number | null
  _getObjects(property: string | NamedNode, options?: GetOptions): SafeClownface<Term, D>
  _create<T extends RdfResource>(term: ResourceNode<D>, mixins?: Mixin[] | [Constructor, ...Mixin[]], options?: ResourceCreationOptions<D>): T & ResourceIndexer
}

export default class RdfResourceImpl<D extends DatasetCore = DatasetCore> implements RdfResource<D> {
  public readonly _selfGraph: SingleContextClownface<ResourceIdentifier, D>
  public readonly _unionGraph: SafeClownface<ResourceIdentifier, D>
  public readonly __initialized: boolean = false
  public readonly _parent?: RdfResource<D>
  private readonly __initializeProperties: (() => boolean)
  public static __ns?: NamespaceBuilder
  public static factory: ResourceFactory = new ResourceFactoryImpl(RdfResourceImpl)
  public static __mixins: Mixin[] = []

  private static _userInitializeProperties(resource: RdfResourceImpl, init: Initializer<RdfResource> = {}): void {
    Object.entries(init)
      .filter(([prop]) => prop !== 'id' && prop !== 'types')
      .forEach(([prop, value]) => {
        (resource as any)[prop] = value
      })

    if (init.types && Array.isArray(init.types)) {
      init.types.forEach(type => resource.types.add(type))
    }
  }

  public constructor(graph: ResourceNode<D>, init: Initializer<any> = {}, parent?: RdfResource<D>) {
    if (graph.term.termType !== 'BlankNode' && graph.term.termType !== 'NamedNode') {
      throw new Error(`RdfResource cannot be initialized from a ${graph.term.termType} node`)
    }

    const selfGraph = cf({
      ...graph,
      term: graph.term,
    })

    if (selfGraph._context[0].graph) {
      this._selfGraph = selfGraph
      this._unionGraph = cf({ dataset: selfGraph.dataset, term: selfGraph.term, graph: undefined })
    } else {
      this._selfGraph = cf({ dataset: selfGraph.dataset, term: selfGraph.term, graph: RDF.defaultGraph() })
      this._unionGraph = cf({ dataset: selfGraph.dataset, term: selfGraph.term })
    }

    this.__initializeProperties = once(() => {
      const self = this as any
      const defaults: Map<string, ObjectOrFactory<any>> = self.constructor.__defaults || new Map<string, ObjectOrFactory<any>>()
      defaults.forEach((value, key) => {
        const currentValue = self[key]
        const valueIsEmptyArray = Array.isArray(currentValue) && currentValue.length === 0
        const valueIsUndefined = typeof currentValue === 'undefined'

        if (!(valueIsEmptyArray || valueIsUndefined)) {
          return
        }

        if (typeof value === 'function') {
          self[key] = value(self)
          return
        }

        self[key] = value
      })

      return true
    })

    this._parent = parent
    this.__initialized = this.__initializeProperties()
    RdfResourceImpl._userInitializeProperties(this, init)
  }

  public get id(): ResourceIdentifier {
    return this._selfGraph.term
  }

  public get _graphId(): Quad_Graph {
    return this._selfGraph._context[0].graph!
  }

  public get types(): TypeCollection<D> {
    return new TypeCollectionCtor(this)
  }

  public get isAnonymous() {
    return this.id.termType === 'BlankNode'
  }

  public hasType(type: string | NamedNode): boolean {
    return this.types.has(type)
  }

  public equals(other: RdfResource | undefined | null): boolean {
    if (!other) {
      return false
    }

    return this.id.equals(other.id)
  }

  public get<T extends RdfResource = RdfResource>(property: string | NamedNode, options?: GetOptions): T | RdfResource<D> | null {
    const objects = this.getArray<T>(property, options)

    if (objects.length > 0) {
      return objects[0]
    }

    return null
  }

  public getArray<T extends RdfResource = RdfResource>(property: string | NamedNode, options?: GetOptions): T[] {
    const values = this._getObjects(property, options)
      .filter(obj => obj.term.termType === 'NamedNode' || obj.term.termType === 'BlankNode')
      .map(obj => {
        return this._create<T>(obj, [], { parent: this })
      })

    if (!values.length) {
      return []
    }

    return values
  }

  public getNumber(property: string | NamedNode): number
  public getNumber(property: string | NamedNode, options?: GetOptions): number | null {
    const [value] = this._getObjects(property, options).toArray()

    if (typeof value === 'undefined') {
      return null
    }

    if (value.term.termType === 'Literal') {
      return parseFloat(value.value)
    }

    throw new Error(`Expected property '${property}' to be a number but found '${value}'`)
  }

  public getString(property: string | NamedNode): string
  public getString(property: string | NamedNode, options?: GetOptions): string | null {
    const [value] = this._getObjects(property, options).toArray()

    if (typeof value === 'undefined') {
      return null
    }

    if (value.term.termType === 'Literal') {
      return value.value
    }

    throw new Error(`Expected property '${property}' to be a literal but found '${value}'`)
  }

  public getBoolean(property: string | NamedNode, options?: GetOptions): boolean {
    const [value] = this._getObjects(property, options).toArray()

    if (typeof value === 'undefined') {
      return false
    }

    if (value.term.termType === 'Literal' && xsd.boolean.equals(value.term.datatype)) {
      return value.term.equals(this._selfGraph.literal(true).term)
    }

    throw new Error(`Expected property '${property}' to be a boolean but found '${value}'`)
  }

  public _getObjects(property: string | NamedNode, { strict }: GetOptions = { strict: false }): SafeClownface<Term, D> {
    const propertyNode = typeof property === 'string' ? this._selfGraph.namedNode(property) : property
    const objects = this._selfGraph.out(propertyNode)

    if (!objects.terms.length && strict) {
      throw new Error(`Value for predicate <${property}> was missing`)
    }

    return objects
  }

  public _create<T extends RdfResource>(term: ResourceNode<D>, mixins?: Mixin[] | [Constructor, ...Mixin[]], options: ResourceCreationOptions<D> = {}): T & ResourceIndexer {
    return (this.constructor as Constructor).factory.createEntity<T>(term, mixins, options)
  }
}

type UserDefinedInterface<T extends RdfResource> = Omit<T, keyof RdfResource>

interface BaseInitializer {
  types?: NamedNode[] | TypeCollection<any>
  id?: RdfResource['id']
}

type InitialNode<Node extends Term = NamedNode | BlankNode> = Node | SingleContextClownface<Node>
type InitialLiteral = InitialNode<Literal>

export type Initializer<T> = Partial<Omit<{
  [P in keyof T]?:
  T[P] extends (infer U)[] ? U extends RdfResource ? Initializer<UserDefinedInterface<U>>[] | InitialNode[]
    : U extends Term ? T[P] | InitialNode<Term>[] : T[P] | InitialLiteral[]
    : T[P] extends RdfResource ? Initializer<UserDefinedInterface<T[P]> & BaseInitializer> | InitialNode :
      T[P] extends Term ? T[P] | InitialNode<T[P]>
        : T[P] | InitialLiteral
}, keyof RdfResource>> & BaseInitializer

type PartialRecursive<T> = T extends object ? { [K in keyof T]?: PartialRecursive<T[K]> } : T

export function fromObject<T extends RdfResource>(initializer: Initializer<UserDefinedInterface<T> & BaseInitializer>): T {
  return initializer as unknown as T
}
