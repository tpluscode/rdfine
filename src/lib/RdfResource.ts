import { Term, NamedNode, DatasetCore, BlankNode } from 'rdf-js'
import cf, { SingleContextClownface } from 'clownface'
import ns from '@rdfjs/namespace'
import once from 'once'

const rdf = ns('http://www.w3.org/1999/02/22-rdf-syntax-ns#')

export interface RdfResource<D extends DatasetCore = DatasetCore> {
  readonly id: Term | null
  readonly types: Term[]
  readonly _node: SingleContextClownface<D>
  hasType (type: string | NamedNode): boolean
}

export default class RdfResourceImpl<D extends DatasetCore = DatasetCore> implements RdfResource<D> {
  public readonly _node: SingleContextClownface<D>
  private readonly __initializeProperties: (() => void)
  public static __ns?: any
  private static __defaults = new Map<string, unknown>()

  public constructor(node: SingleContextClownface<D> | { dataset: D; term: NamedNode | BlankNode; graph?: NamedNode }) {
    const contexts = cf(node).toArray()

    if (contexts.length !== 1) {
      throw new Error('RdfResource can only be initialized from a single node. Got ' + contexts.length)
    }

    this._node = cf(node).toArray()[0]

    this.__initializeProperties = once(() => {
      const self = this as any
      RdfResourceImpl.__defaults.forEach((value, key) => {
        if (typeof self[key] === 'undefined') {
          self[key] = value
        }
      })
    })

    this.__initializeProperties()
  }

  public get id() {
    if (!this._node.term) throw new Error('Graph context does not represent a single node')

    return this._node.term
  }

  public get types() {
    return this._node.out(rdf.type).terms
  }

  public hasType(type: string | NamedNode) {
    const typeNode = typeof type === 'string' ? this._node.namedNode(type) : type

    return this._node.has(rdf.type, typeNode).terms.length > 0
  }
}
