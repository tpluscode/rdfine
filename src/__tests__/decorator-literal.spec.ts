import { prefixes } from '@zazuko/rdf-vocabularies'
import { property } from '..'
import RdfResource from '../lib/RdfResource'
import { parse, vocabs } from './_helpers'
import { Literal } from 'rdf-js'
import { literal } from 'rdf-data-model'
import rdfExt from 'rdf-ext'

const { ex, schema } = vocabs

describe('decorator', () => {
  describe('literal', () => {
    describe('getter', () => {
      it('returns string value of literal', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res schema:name "John Doe" .
        `)
        class Resource extends RdfResource {
          @property.literal({ path: schema.name })
          name?: string
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.name).toEqual('John Doe')
      })

      it('returns boolean when type is set', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          @prefix xsd: <${prefixes.xsd}> .
          
          ex:res ex:isMarried true .
          ex:res ex:isSingle "false"^^xsd:boolean .
        `)
        class Resource extends RdfResource {
          @property.literal({ path: ex.isMarried, type: Boolean })
          married?: boolean

          @property.literal({ path: ex.isSingle, type: Boolean })
          single?: boolean
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.single).toStrictEqual(false)
        expect(instance.married).toStrictEqual(true)
      })

      it('returns integer when type is set', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          @prefix xsd: <${prefixes.xsd}> .
          
          ex:res ex:age "30" .
        `)
        class Resource extends RdfResource {
          @property.literal({ path: ex.age, type: Number })
          age!: number
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.age).toStrictEqual(30)
      })

      it('returns float when type is set and node is floating point', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          @prefix xsd: <${prefixes.xsd}> .
          
          ex:res ex:age 30.6 .
        `)
        class Resource extends RdfResource {
          @property.literal({ path: ex.age, type: Number })
          age!: number
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.age).toStrictEqual(30.6)
      })
    })

    describe('setter', () => {
      it('accepts raw term', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          @prefix xsd: <${prefixes.xsd}> .
          
          ex:res schema:name "Johann"@de .
        `)
        class Resource extends RdfResource {
          @property.literal({ path: schema.name })
          name?: string | Literal
        }

        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // when
        instance.name = literal('John', 'en-gb')

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('replaces boolean object value', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          @prefix xsd: <${prefixes.xsd}> .
          
          ex:res ex:isMarried true .
        `)

        class Resource extends RdfResource {
          @property.literal({ path: ex.isMarried, type: Boolean })
          married?: boolean
        }

        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // when
        instance.married = false

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('sets xsd:integer literal for int number', async () => {
        // given
        const dataset = rdfExt.dataset()

        class Resource extends RdfResource {
          @property.literal({ path: ex.age, type: Number })
          age!: number
        }

        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // when
        instance.age = 30

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('sets xsd:float literal for float number', async () => {
        // given
        const dataset = rdfExt.dataset()

        class Resource extends RdfResource {
          @property.literal({ path: ex.age, type: Number })
          age!: number
        }

        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // when
        instance.age = 30.4

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('replaces string object value', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          @prefix xsd: <${prefixes.xsd}> .
          
          ex:res schema:name "John" .
        `)

        class Resource extends RdfResource {
          @property.literal({ path: schema.name })
          name?: string
        }

        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // when
        instance.name = 'Jane'

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('unsets string object when null is set', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          @prefix xsd: <${prefixes.xsd}> .
          
          ex:res schema:name "John" .
        `)

        class Resource extends RdfResource {
          @property.literal({ path: schema.name })
          name: string | null = null
        }

        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // when
        instance.name = null

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })
    })

    describe('initial', () => {
      it('sets default value when not present in dataset', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property.literal({
            path: schema.name,
            initial: 'foo',
          })
          name = 'foo'
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.name).toEqual('foo')
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('does not set initialized value when present in initial dataset', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          @prefix xsd: <${prefixes.xsd}> .
          
          ex:res schema:name "bar" .
        `)
        class Resource extends RdfResource {
          @property.literal({
            path: schema.name,
            initial: 'foo',
          })
          name!: string
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.name).toEqual('bar')
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('sets initial value from node', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property.literal({
            path: schema.name,
            initial: literal('foo'),
          })
          name!: string
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.name).toEqual('foo')
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('sets initial value from function', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property.literal({
            path: schema.name,
            initial: (self: Resource) => self.id.value,
          })
          name!: string
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.name).toEqual(instance.id.value)
        expect(dataset.toCanonical()).toMatchSnapshot()
      })
    })
  })
})
