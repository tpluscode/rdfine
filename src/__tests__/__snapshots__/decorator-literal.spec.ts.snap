// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorator literal initial does not set initialized value when present in initial dataset 1`] = `
"<http://example.com/res> <http://schema.org/name> \\"bar\\" .
"
`;

exports[`decorator literal initial sets default value when not present in dataset 1`] = `
"<http://example.com/res> <http://schema.org/name> \\"foo\\" .
"
`;

exports[`decorator literal initial sets initial value from function 1`] = `
"<http://example.com/res> <http://schema.org/name> \\"http://example.com/res\\" .
"
`;

exports[`decorator literal initial sets initial value from node 1`] = `
"<http://example.com/res> <http://schema.org/name> \\"foo\\" .
"
`;

exports[`decorator literal setter accepts raw term 1`] = `
"<http://example.com/res> <http://schema.org/name> \\"John\\"@en-gb .
"
`;

exports[`decorator literal setter replaces boolean object value 1`] = `
"<http://example.com/res> <http://example.com/isMarried> \\"false\\"^^<http://www.w3.org/2001/XMLSchema#boolean> .
"
`;

exports[`decorator literal setter replaces string literal node value 1`] = `
"<http://example.com/res> <http://schema.org/name> \\"Jane\\"^^<http://example.com/Name> .
"
`;

exports[`decorator literal setter replaces string literal node value 2`] = `
"<http://example.com/res> <http://schema.org/name> \\"Jane\\"^^<http://example.com/Name> .
"
`;

exports[`decorator literal setter replaces string object value 1`] = `
"<http://example.com/res> <http://schema.org/name> \\"Jane\\" .
"
`;

exports[`decorator literal setter sets xsd:float literal for float number 1`] = `
"<http://example.com/res> <http://example.com/age> \\"30.4\\"^^<http://www.w3.org/2001/XMLSchema#float> .
"
`;

exports[`decorator literal setter sets xsd:integer literal for int number 1`] = `
"<http://example.com/res> <http://example.com/age> \\"30\\"^^<http://www.w3.org/2001/XMLSchema#integer> .
"
`;

exports[`decorator literal setter sets xsd:long literal for bigint 1`] = `
"<http://example.com/res> <http://example.com/age> \\"9007199254740991\\"^^<http://www.w3.org/2001/XMLSchema#long> .
"
`;

exports[`decorator literal setter unsets string object when null is set 1`] = `""`;
